rust/src/mqtt/parser.rs

```diff
@@ -106,13 +106,13 @@ fn parse_properties(input: &[u8], precond: bool) -> IResult<&[u8], Option<Vec<MQ
     // parse properties length
     match parse_mqtt_variable_integer(input) {
         Ok((rem, mut proplen)) => {
-            if proplen == 0 {
+            if proplen == 0 || proplen as usize > rem.len() {
```

Comment:
in what condition can this happen? Should it lead to error handling and/or setting of event?

Reply:
In theory, one can craft a message that has this value set to a larger value than the total length specified in the header. One could imagine raising an event here, yes.

Reply:
@satta It looks like there are more problems with MQTT :
if `parse_connect` returns `Err::Incomplete` (for example because, there is not enough bytes for the properties), we will keep buffering forever because we have the full message with its length already cf `let rem = &fullrem[..len];`

Am I correct ?

I guess the solution is to use nom's `complete` (which turns Err::Incomplete into Err:Error)

Then, I guess there is more work to do to get the error to create an app-layer event...

Reply:
I also think we can have `parse_properties` more idiomatic with nom by using `take` and `many`

Reply:
> @satta It looks like there are more problems with MQTT : if `parse_connect` returns `Err::Incomplete` (for example because, there is not enough bytes for the properties), we will keep buffering forever because we have the full message with its length already cf `let rem = &fullrem[..len];`
> 
> Am I correct ?

Hmm, I think you are right. We make sure we have the announced full length of the message (as defined in the common header `header.remaining_length`) buffered already because we try to handle buffering before trying to do any type-specific parsing in line 509ff:
```
            // We have not exceeded the maximum length limit, but still do not
            // have enough data in the input buffer to handle the full
            // message. Signal this by returning an Incomplete IResult value.
            if fullrem.len() < len {
                return Err(Err::Incomplete(Needed::Size(len - fullrem.len())));
            }
```
So if we notice later that the pre-filled buffer is not enough, we have a problem. But, maybe it helps to know that in a valid control packet ("message") we should _never_ generate `Err::Incomplete` somewhere else than in the code above since we SHOULD have buffered all the expected data to parse the whole message already. So any mismatch of expected and actually matched length will be an indication of an inconsistency in the message, and we should simply notify this and stop parsing since obviously something's amiss.

> I guess the solution is to use nom's `complete` (which turns Err::Incomplete into Err:Error)

I guess so, too. And ideally this should be done globally, whenever one of the `parse_*()` functions wants more data. What do you think?

Reply:
Created https://redmine.openinfosecfoundation.org/issues/5018 about this

A global solution is good.
That would be wrapping the 
```
match message_type {
                MQTTTypeCode::CONNECT => match parse_connect(rem) {
```
in a `nom::complete` function

Let me know if you need more help ;-)

Reply:
Something like this? -> https://github.com/satta/suricata/commit/a3d26d59aa3ab25f4a16ee30be2075b0abdd41eb

Reply:
Not sure about the `move` but the `complete` look nice :-)
Waiting for the PR ;-)

Reply:
Regarding the `move`: Got that from https://github.com/Geal/nom/blob/main/doc/upgrading_to_nom_5.md
Will try to get a suricata-verify case going before starting the PR.

Reply:
See https://github.com/OISF/suricata/pull/6988

Reply:
> From what I could see, the original issue should not be a problem at all in the current master code, since nom7::bytes::complete::take() is used to pre-fetch a subslice of length proplen in line 116 of rust/src/mqtt/parser.rs

Hey @satta took me some time to get back to this

So, you say my patch from https://github.com/OISF/suricata/pull/6890/commits/3c265611fe830b5e7e4f395d762ba223bd316423 was enough ?

I do not think it is enough for all cases.
For instance `parse_connect` calls`parse_mqtt_string` which calls nom `length_data` which can return incomplete cf  https://docs.rs/nom/latest/nom/multi/fn.length_data.html
What do you think ?

Reply:
I think that the current solution wrapping everything in `complete()` also catches that and should be preferred. See my latest MR.

Reply:
discussion continued in #7009 :-)